docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1058 @@
+| 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 |
+-----------------------------------------------------------------
+```
+
+`Vector256<T>` is twice as big as `Vector128<T>`, so when it is hardware accelerated, and the data is large enough we should prefer it over a `Vector128<T>`. To check the acceleration, we need to use `Vector128.IsHardwareAccelerated` and `Vector256.IsHardwareAccelerated` properties.
+
+We also must account for the size of the input. It needs to be at least of the size of a single vector to be able to execute the vectorized code path. `Vector128<T>.Count` and `Vector256<T>.Count` return the size of a vector of given type in bytes.
+Both APIs are turned into constants (no method call is required to retrieve the information) by the Just-In-Time compiler. It's not true for pre-compiled code (NativeAOT).
+
+That is why the code is very often structured like this:
+
+```cs
+void CodeStructure(ReadOnlySpan<byte> buffer)
+{
+    if (Vector256.IsHardwareAccelerated && buffer.Length >= Vector256<byte>.Count)
+    {
+        // Vector256 code path
+    }
+    else if (Vector128.IsHardwareAccelerated && buffer.Length >= Vector128<byte>.Count)
+    {
+        // Vector128 code path
+    }
+    else
+    {
+        // non-vectorized && small inputs code path
+    }
+}
+```
+
+**Both vector types provide almost identical features**, but arm64 hardware does not support `Vector256` yet, so for the sake of simplicity we will be using `Vector128` in all examples and assuming **little endian** architecture. Which means that all examples used in this document assume that they are being executed as part of the following `if` block:
+
+```cs
+else if (Vector128.IsHardwareAccelerated && buffer.Length >= Vector128<byte>.Count)
+{
+    // Vector128 code path
+}
+```
+
+### Testing
+
+Such a code structure requires us to **test all possible code paths**:
+
+* `Vector256` is accelerated:
+  * The input is large enough to benefit from vectorization with `Vector256`.
+  * The input is not large enough to benefit from vectorization with `Vector256`, but it can benefit from vectorization with `Vector128` (when `Vector256` is accelerated then `Vector128` and smaller vectors are also).
+  * The input is too small to benefit from any kind of vectorization.
+* `Vector128` is accelerated
+  * The input is large enough to benefit from vectorization with `Vector128`.
+  * The input is too small to benefit from any kind of vectorization.
+* Neither `Vector128` or  `Vector256` are accelerated.
+
+It's possible to implement tests that cover some of the scenarios based on the size, but it's impossible to toggle hardware acceleration from unit test level. It can be controlled with environment variables before .NET process is started:
+
+* When `DOTNET_EnableAVX2` is set to `0`, `Vector256.IsHardwareAccelerated` returns `false`.
+* When `DOTNET_EnableAVX` is set to `0`, `Vector128.IsHardwareAccelerated` returns `false`.
```

Comment:
A typo? EnableAVX=0 should not affect `Vector128.IsHardwareAccelerated`

Reply:
@EgorBo great catch! I was a bug, I was convinced that it was true ;)